[master]Add Ballerina Deployment Guide (#9621)
ballerina-platform/ballerina-dev-website: swan-lake/development-tutorials/deployment-guide/serverless-deployment.md

---
layout: ballerina-cloud-left-nav-pages-swanlake
title: Serverless deployment
description: Serverless architecture allows developers to focus on writing application logic without worrying about managing servers or infrastructure. AWS Lambda automatically handles provisioning, scaling, and maintaining resources, letting you concentrate solely on your code. Following are the  Ballerina functions in a serverless environment, 
keywords: ballerina, programming language, services, cloud, kubernetes, docker
active: ballerina-deployment-guideLines
intro: Serverless architecture allows developers to focus on writing application logic without worrying about managing servers or infrastructure. AWS Lambda automatically handles provisioning, scaling, and maintaining resources, letting you concentrate solely on your code. Following are the  Ballerina functions in a serverless environment, 
---

1. [AWS Lambda](#aws-lambda)

> 2. [Azure Functions](#azure-function)

Reviewer: Wrong link

## AWS Lambda

In this guide, we'll walk through deploying a serverless application using AWS Lambda with Ballerina. You'll learn how to set up, deploy, and manage your application, simplifying your production workflow and eliminating the need to manage servers.

> An AWS Lambda function can be triggered by various AWS services. You can find the list of supported notification types below.

Reviewer: ```suggestion
An AWS Lambda function can be triggered by various AWS services. You can find the list of supported triggers  below.
```

- Direct invocation
- Simple Queue Service (SQS)
- Simple Storage Service (S3)
- DynamoDB 
- Simple Email Service (SES)

> - API Gateway

Reviewer: Add the links similar to https://ballerina.io/learn/aws-lambda/#supported-triggers

Reviewer: Btw do we need the above list here as it is repeated in https://ballerina.io/learn/aws-lambda/ as well?

Author: Removed repititve content.

This example demonstrates how to write a simple echo function in AWS Lambda.

> For more information, see [the AWS Lambda learn guide](https://ballerina.io/learn/aws-lambda/).

Reviewer: Now we don't have a learn section tile for this [aws-lambda ](https://ballerina.io/learn/aws-lambda/)page and [azure-functions](https://ballerina.io/learn/azure-functions/) page. How this will impact the left navigation when the reader lands on such a page? Better to include them under Serverless deployment in left nav. 

And also can we include next level titles in the landing page tile as well as indented links under serverless deployment. Adding @sm1990 as well.

<img width="405" alt="image" src="https://github.com/user-attachments/assets/11d9981e-222d-45cf-b013-3a7abf02c117">

Reviewer 2: @anupama-pathirage & @lnash94 
> Now we don't have a learn section tile for this [aws-lambda ](https://ballerina.io/learn/aws-lambda/)page and [azure-functions](https://ballerina.io/learn/azure-functions/) page. How this will impact the left navigation when the reader lands on such a page? Better to include them under Serverless deployment in left nav.

We can create a new section called `Serverless deployment` under `Deployment guidelines`. 
Once you click on `Serverless deployment` in left nav, it will expand the section in the left nav and show 3 pages. 

- Serverless deployment overview
- AWS
- Azure

> And also can we include next level titles in the landing page tile as well as indented links under serverless deployment.

We are currently showing headings/titles only up to the 3rd level in the `Learn` landing page. If we are going deeper, we'll have to update other sections as well.

> 1. Set up the prerequisites 

Reviewer: ```suggestion
1. Set up the prerequisites.
```

   For the instructions, see [Set up the prerequisites](https://ballerina.io/learn/aws-lambda/#set-up-the-prerequisites)

> 2. Here is an example of a Ballerina code for the function

Reviewer: ```suggestion
2. Here is an example of a Ballerina code for the function.
```

```
$ aws lambda create-function --function-name echo --zip-file fileb://aws-ballerina-lambda-functions.zip --handler aws-lambda-hello-world.echo --runtime provided --role arn:aws:iam::908363916111:role/lambda-role--layers arn:aws:lambda:us-west-1:134633749276:layer:ballerina-jre11:6 --memory-size 512 --timeout 10
```
5. Execute the commands below to invoke the function.

```
$ echo '{"MESSAGE":"HELLO"}' > input.json
$ aws lambda invoke --function-name echo --payload fileb://input.json echo-response.txt
{
"ExecutedVersion": "$LATEST",
"StatusCode": 200
}
$ cat echo-response.txt{"MESSAGE":"HELLO"}

```
More examples can be found here,
- [AWS Lambda - Execution context](https://ballerina.io/learn/by-example/aws-lambda-execution-context/)
- [AWS Lambda - S3 trigger](https://ballerina.io/learn/by-example/aws-lambda-s3-trigger/)
- [AWS Lambda - DynamoDB trigger](https://ballerina.io/learn/by-example/aws-lambda-dynamodb-trigger/)

### CI/CD deployment with AWS Lambda

This GitHub Action workflow automates the continuous integration and continuous deployment (CI/CD) process for a Ballerina project. 

> It triggers every push to the repository, and builds the project.

Reviewer: same as https://github.com/ballerina-platform/ballerina-dev-website/pull/9621/files#r1878558376

          timeout: 30
          memory_size: 512

      # Optional: Test the Lambda Function
      - name: Invoke Lambda Function
        run: |
          aws lambda invoke --function-name your_lambda_function_name --payload '{}' response.json
          cat response.json
        env:
          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
          AWS_REGION: ${{ secrets.AWS_REGION }}
```
- **Step 01:** The push event can initiate a build and deployment process
- **Step 02:** This command tells Ballerina to compile the project specifically for AWS Lambda. It optimizes the output for Lambda’s serverless runtime requirements.
- **Step 03:** Package the compiled Ballerina application into a ZIP file suitable for AWS Lambda deployment.
- **Step 04:** Ensures that the workflow has the necessary credentials to interact with AWS services.
- **Step 05:** Deploy the packaged Ballerina function to AWS Lambda by uploading the ZIP file.
- **Step 06:** Test the deployed Lambda function by invoking it and outputting the response to confirm that the function works as expected.

> See [AWS Lambda](https://ballerina.io/learn/aws-lambda/) for more information.

Reviewer: This is referred from multiple place and burried inside the text. So can we add the aws-lambda guide and azure-funtion guide links to the bottom of this page and that will help to find the links easily

## Azure Functions

In this guide, we'll walk through deploying a serverless application using Azure Functions with Ballerina. 
You'll learn how to set up, deploy, and manage your application, leveraging the power of Azure's serverless platform 
to simplify your workflows and eliminate the overhead of server management.

This example demonstrates how to write a simple echo function in Azure Functions using Ballerina, where triggers 
are represented by listeners attaching an `af:HttpListener` to a service that implies an HTTP trigger. 
The resource method behaves like a `ballerina/http` service, In the provided code sample, there's an empty service path and a resource path
named `hello` with a `get` accessor expecting a `name` query parameter, and the `ballerinax/azure_functions` package automatically handles the required artifact generation and data binding.

For more information, see [the Azure deployment guide](https://ballerina.io/learn/azure-functions/).

1. Set up the prerequisites

> For instructions, see [Set up the prerequisites.](https://ballerina.io/learn/azure-functions/#set-up-the-prerequisites)

Reviewer: Check the above two lines, appear in single line and not making sense.

Author: Aligned it to one sentence

> 2. Here is an example of a Ballerina code for the function

Reviewer: ```suggestion
2. Here is an example Ballerina code.
```

```ballerina
import ballerinax/azure.functions;

// This function gets triggered by an HTTP call with the name query parameter and 
// returns a processed HTTP output to the caller.
service / on new functions:HttpListener() {
    resource function get hello(string name) returns string {
        return string `Hello,  ${name}!`;
    }
}
```

> 3. Use `bal build` for generating the Azure Function artifacts.

Reviewer: ```suggestion
3. Use `bal build` to generate the Azure Function artifacts.
```

```
$ bal build --cloud="azure_functions"
```

> 4. Execute the Azure CLI command given by the compiler to create and publish the functions by replacing <function_app_name> with your respective function app name.

Reviewer: ```suggestion
4. Execute the Azure CLI command given by the compiler to create and publish the functions by replacing `<function_app_name>` with your respective function app name.
```

```
$ func azure functionapp publish <function_app_name> --script-root target/azure_functions
```
5. Execute the command below to invoke the function by replacing <function_app_name> with your respective function app name.
```
$ curl https://<function_app_name>.azurewebsites.net/hello\?name\=Jack
```

### CI/CD with Azure Function

This GitHub Action workflow automates the continuous integration and continuous deployment (CI/CD) process for a Ballerina project.

> It triggers every push to the repository, and builds the project.

Reviewer: Shouldn't this be ` It is triggered by every push to the repository and automatically builds the project.`?

Author: updated

      - name: Set up Azure Functions Core Tools
        uses: philippelemaire/install-azure-functions-core-tools@v1

      # Step 4: Deploy to Azure Functions
      - name: Azure Functions Action
        uses: Azure/functions-action@v1
        with:
          app-name: 'your-function-app-name'
          package: './'
        env:
          AZURE_FUNCTIONAPP_PUBLISH_PROFILE: ${{ secrets.AZURE_FUNCTIONAPP_PUBLISH_PROFILE }}

      # Optional: Test the Azure Function
      - name: Invoke Azure Function
        run: |
          curl https://your-function-app-name.azurewebsites.net/api/your-function-name
```
- **Step 01:** The push event can initiate a build and deployment process
- **Step 02:** This step compiles the Ballerina application to generate an executable JAR file. The `bal build` command compiles the code and packages it, making it ready for deployment.
- **Step 03:** Install the Azure Functions Core Tools, which are necessary for deploying and testing Azure Functions in this GitHub Actions environment.
- **Step 04:** Deploy the built Ballerina application to an Azure Function App. This step uploads the packaged function to Azure and configures it for serverless execution.

> See [Azure Functions](https://ballerina.io/learn/azure-functions/) for information about deployment options can be used in Azure Functions.

Author: See [Azure Functions](https://ballerina.io/learn/azure-functions/) for information about deployment options that can be used in Azure Functions.
